drivers/storage/portworx/portworx.go

```diff
@@ -189,114 +189,12 @@ func (p *portworx) SetDefaultsOnStorageCluster(toUpdate *corev1.StorageCluster)
-		}
-
-		if pxutil.IsCSIEnabled(toUpdate) {
-			if toUpdate.Status.DesiredImages.CSIProvisioner == "" ||
-				pxVersionChanged ||
-				autoUpdateComponents(toUpdate) {
-				toUpdate.Status.DesiredImages.CSIProvisioner = release.Components.CSIProvisioner
-				toUpdate.Status.DesiredImages.CSINodeDriverRegistrar = release.Components.CSINodeDriverRegistrar
-				toUpdate.Status.DesiredImages.CSIDriverRegistrar = release.Components.CSIDriverRegistrar
-				toUpdate.Status.DesiredImages.CSIAttacher = release.Components.CSIAttacher
-				toUpdate.Status.DesiredImages.CSIResizer = release.Components.CSIResizer
-				toUpdate.Status.DesiredImages.CSISnapshotter = release.Components.CSISnapshotter
-				toUpdate.Status.DesiredImages.CSIHealthMonitorController = release.Components.CSIHealthMonitorController
-			}
-			if autoUpdateCSISnapshotController(toUpdate) &&
-				(toUpdate.Status.DesiredImages.CSISnapshotController == "" ||
-					pxVersionChanged ||
-					autoUpdateComponents(toUpdate)) {
-				toUpdate.Status.DesiredImages.CSISnapshotController = release.Components.CSISnapshotController
-			}
-		}
-
-		if toUpdate.Spec.Monitoring != nil && toUpdate.Spec.Monitoring.Prometheus != nil {
-			prometheusVersionChanged := p.hasPrometheusVersionChanged(toUpdate)
-			if toUpdate.Spec.Monitoring.Prometheus.Enabled &&
-				(toUpdate.Status.DesiredImages.PrometheusOperator == "" || pxVersionChanged || prometheusVersionChanged) {
-				toUpdate.Status.DesiredImages.Prometheus = release.Components.Prometheus
-				toUpdate.Status.DesiredImages.PrometheusOperator = release.Components.PrometheusOperator
-				toUpdate.Status.DesiredImages.PrometheusConfigMapReload = release.Components.PrometheusConfigMapReload
-				toUpdate.Status.DesiredImages.PrometheusConfigReloader = release.Components.PrometheusConfigReloader
-			}
-			if toUpdate.Spec.Monitoring.Prometheus.AlertManager != nil &&
-				toUpdate.Spec.Monitoring.Prometheus.AlertManager.Enabled &&
-				(toUpdate.Status.DesiredImages.AlertManager == "" || pxVersionChanged || prometheusVersionChanged) {
-				toUpdate.Status.DesiredImages.AlertManager = release.Components.AlertManager
-			}
-		}
-
-		// Reset the component update strategy if it is 'Once', so that we don't
-		// upgrade components again during next reconcile loop
-		if toUpdate.Spec.AutoUpdateComponents != nil &&
-			*toUpdate.Spec.AutoUpdateComponents == corev1.OnceAutoUpdate {
-			toUpdate.Spec.AutoUpdateComponents = nil
+		// The version from manifest does not match desired version, this may be caused by wrong configMap version.
+		if release.PortworxVersion != "" && toUpdate.Spec.Version != "" && release.PortworxVersion != toUpdate.Spec.Version {
+			msg := fmt.Sprintf("Version in StorageCluster %s does not match version manifest %s, please edit StorageCluster or px-versions configMap.", toUpdate.Spec.Version, release.PortworxVersion)
+			p.warningEvent(toUpdate, util.FailedComponentReason, msg)
```

Comment:
if we hit this, do we still proceed? if we don't update desired images in this case, we should not reset them either I feel.

Reply:
do you mean we should not reset to empty or something else?

Reply:
nvm so if the versions don't match, we only avoid updating component images, then everything else should still work fine with the old desired images.